keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1462,225 +1279,309 @@ prevented. Deleting the entire cluster also leaves resources allocated outside
-    // resources have not been allocated yet.
+    // Shareable indicates whether the allocated claim is meant to be shareable
+    // by multiple consumers at the same time.
     // +optional
-    Allocation *AllocationResult
+    Shareable bool
+}
 
-    // ReservedFor indicates which entities are currently allowed to use
-    // the claim. A Pod which references a ResourceClaim which is not
-    // reserved for that Pod will not be started.
+// Constraint must have one and only one field set.
+type Constraint struct {
+    // All devices must have this attribute and its value must be the same.
+    //
+    // For example, if you specified "numa.dra.example.com" (a hypothetical example!),
+    // then only devices in the same NUMA node will be chosen.
     //
-    // There can be at most 32 such reservations. This may get increased in
-    // the future, but not reduced.
     // +optional
-    ReservedFor []ResourceClaimConsumerReference
+    // +listType=atomic
+    MatchAttribute *string
+
+    // Future extension, not part of the current design:
+    // A CEL expression which compares different devices and returns
+    // true if they match.
+    //
+    // Because it would be part of a one-of, old schedulers will not
+    // accidentally ignore this additional, for them unknown match
+    // criteria.
+    //
+    // matcher string
 }
 
-// ReservedForMaxSize is the maximum number of entries in
-// claim.status.reservedFor.
-const ResourceClaimReservedForMaxSize = 32
-```
+// Request is a request for one of many resources required for a claim.
+// This is typically a request for a single resource like a device, but can
+// also ask for one of several different alternatives.
+type Request struct {
+    // The name can be used to reference this request in a pod.spec.containers[].resources.claims
+    // entry.
+    //
+    // Must be a DNS label.
+    Name string
 
-##### ResourceClaimParameters
+    *ResourceRequestDetail
 
-```go
-// ResourceClaimParameters defines resource requests for a ResourceClaim in an
-// in-tree format understood by Kubernetes.
-type ResourceClaimParameters struct {
-    metav1.TypeMeta
-    // Standard object metadata
-    metav1.ObjectMeta
+    // FUTURE EXTENSION:
+    //
+    // OneOf contains a list of requests, only one of which must be satisfied.
+    // Requests are listed in order of priority.
+    //
+    // +optional
+    // +listType=atomic
+    // OneOf []ResourceRequestDetail
+}
 
-    // If this object was created from some other resource, then this links
-    // back to that resource. This field is used to find the in-tree representation
-    // of the claim parameters when the parameter reference of the claim refers
-    // to some unknown type.
-    GeneratedFrom *ResourceClaimParametersReference
+type ResourceRequestDetail struct {
+    // When referencing a DeviceClass, a request inherits additional
+    // configuration and requirements.
+    //
+    // +optional
+    DeviceClassName *string
```

Comment:
I think the common case for a class will be to preselect the driver that a device is associated with. The driver implementor can provide such a class as part of the helm chart deployment for the driver.

So my only concern with requiring a class is when we want to allow selection of a device from different drivers. No single driver could provide a class to cover this use-case. The admin would need to craft a custom class that OR'd the possible drivers that the device can be selected from.

Reply:
Yes, though that case we can handle for example, by having a standard attribute that the class can select on, and having a class that ships with k8s directly. For example, an attribute `resource.k8s.io/nic_type` with values `raw`, `sriov-pf`, `sriov-vf` or something. We can figure that out later.

Reply:
> "and using it for quota"

Let's discuss that in https://github.com/kubernetes-sigs/wg-device-management/issues/24. Doing it in this PR will not be focused enough.

In that issue, I gave several examples why quota by class at admission time is not working.

> The admin would need to craft a custom class that OR'd the possible drivers that the device can be selected from.

To me, this is the biggest drawback. We create more work for admins and users and make app deployments non-portable across clusters because class names are not standardized. If it worked, it would at least be worth it, but I don't think it does.

Reply:
I still think we should make class required; we can relax it in the future if we find it's not good. I don't agree the OR'd drivers will be needed, as we can solve the "cross driver class" issue with "standard attributes".

As you said, we can debate quota implementation elsewhere, but if we make this required now we don't bo ourselves in.

Reply:
I can live with making a class required to buy us time, but implementing a quota system based on that needs further discussion.

Reply:
Ok, let's do that. thanks

Reply:
I've thought a bit more about it and I don't buy the argument anymore that `DeviceClassName` has to be required for everyone just because there might be an admission mechanism that depends on that.

Suppose that, hypothetically speaking, someone defined and implemented such an admission check. Then that same check can enforce that the `DeviceClass` field is set in namespaces where the check is enabled. It can remain optional if no such check gets implemented or if the namespace configuration doesn't enforce class-based quotas. Whether the field has to be set becomes part of the documentation of the cluster, together with the corresponding documentation that tells users which classes are available to them.

That gives us the best of both worlds:
- Class-based quotas can be added later or downstream with an admission webhook.
- There  is no negative effect on admins or users which don't want the mechanism.

Allowing the creation of a claim with no `DeviceClassName` before such a mechanism gets implemented falls into the same category as "admitting something before a quota gets defined or changed": it was allowed in the past and thus remains allowed. In practice, such a mechanism would probably be in place before a cluster is handed over to users.

Reply:
Quota is only one reason. Other reasons:
- Consistency with storage and other areas (though not all classes are required)
- Users won't necessarily know about drivers, etc.  - class provides a way to abstract that into specific device qualities that a user understands
- I just don't think a claim without a class makes much sense. It requires a lot of user knowledge to create a meaningful claim

Given that it's easier to relax the constraint than make it required when it wasn't, it seems safest to make it required now.

Reply:
> Consistency with storage

We already diverged. I wouldn't say that consistency for consistencies sake should be a goal. Instead, we need to consider the pros and cons of classes in our design.

> Users won't necessarily know about drivers

That's an argument for encouraging the use of classes by vendors and in documentation. I don't disagree.

But IMHO this does not support the statement that classes must be used.

> It requires a lot of user knowledge to create a meaningful claim

Isn't that the same argument ("classes are easier")?

> Given that it's easier to relax the constraint than make it required when it wasn't

I don't agree that this matters for an alpha release. We could flip from optional to required during the promotion from alpha to beta. It becomes harder later.

Quite the opposite: by making it optional in the alpha, people can experiment with both approaches.

Reply:
I added an `UNRESOLVED` below as reminder that we need to re-evaluate this before beta at the latest.

@johnbelamaric: as you are the proponent of making it required, I assigned that one to you in the text.

Reply:
SGTM